File: apps/desktop/src/lib/ai.ts
import {
  LOCAL_AI_BASE_URL,
  type AskMessage,
  type AskNoteSummary,
  type AskResponse,
  type AudioSessionChunkResponse,
  type AudioSessionCreateResponse,
  type AudioSessionFinalizeRequest,
  type AudioSessionFinalizeResponse,
  type ClassOverviewRequest,
  type ClassOverviewResponse,
  type FlashcardsResponse,
  type QuizResponse,
  type StudyPlanResponse,
  type SummaryResponse,
} from "@studynest/shared";

async function localBase(): Promise<string> {
  try {
    if (typeof window !== "undefined" && window.studynest?.sidecarBaseUrl) {
      const url = await window.studynest.sidecarBaseUrl();
      if (url?.trim()) return url.trim();
    }
  } catch {
    /* preload not ready yet */
  }
  return LOCAL_AI_BASE_URL;
}

/** Desktop uses only the bundled local assistant — no remote AI fallback. */
async function localOnly<T>(path: string, body: unknown): Promise<T> {
  const base = await localBase();
  const res = await fetch(`${base}${path}`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(120_000),
  });
  if (!res.ok) throw new Error(`local ${path} ${res.status}`);
  return (await res.json()) as T;
}

export const ai = {
  async summarize(args: { note_id: string; title: string; content: string }) {
    return localOnly<SummaryResponse>("/local-ai/summarize", args);
  },
  async flashcards(args: {
    note_id: string;
    title: string;
    content: string;
    count?: number;
  }) {
    return localOnly<FlashcardsResponse>("/local-ai/flashcards", {
      count: 8,
      ...args,
    });
  },
  async quiz(args: {
    note_id: string;
    title: string;
    content: string;
    count?: number;
    /** Restricts question types — server may ignore unknown values. */
    types?: Array<"multiple_choice" | "true_false" | "short_answer">;
  }) {
    return localOnly<QuizResponse>("/local-ai/quiz", { count: 5, ...args });
  },
  async studyPlan(args: {
    goal: string;
    exam_date?: string | null;
    notes: Array<{ id: string; title: string; summary?: string | null }>;
    days_available?: number;
  }) {
    return localOnly<StudyPlanResponse>("/local-ai/study-plan", args);
  },
  async simpleExplain(args: { note_id: string; title: string; content: string }) {
    return localOnly<SummaryResponse>("/local-ai/simple-explain", args);
  },
  async ask(args: {
    class_name: string;
    class_subtitle?: string | null;
    recent_notes: AskNoteSummary[];
    weak_topics: string[];
    history: AskMessage[];
    question: string;
  }) {
    return localOnly<AskResponse>("/local-ai/ask", args);
  },
  async classOverview(args: ClassOverviewRequest) {
    return localOnly<ClassOverviewResponse>("/local-ai/class-overview", args);
  },

  /* ----- Chunked audio → notes (Gemma 4 E4B) ----- */

  /** Open a new audio session. The response tells us whether the
   * sidecar will use the real Gemma 4 backend or its placeholder stub
   * — surfaced in the toast so the user isn't lied to. */
  async createAudioSession(): Promise<AudioSessionCreateResponse> {
    const base = await localBase();
    const res = await fetch(`${base}/local-ai/audio-session`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: "{}",
    });
    if (!res.ok) throw new Error(`createAudioSession ${res.status}`);
    return (await res.json()) as AudioSessionCreateResponse;
  },

  /** Append one ≤30s WAV chunk in order. Caller MUST await each
   * promise before sending the next chunk so the server's per-session
   * state stays in arrival order. */
  async appendAudioChunk(args: {
    sessionId: string;
    chunk: Blob;
    index: number;
    total: number;
    /** Optional abort wiring — used by the toast's "Cancel" button. */
    signal?: AbortSignal;
  }): Promise<AudioSessionChunkResponse> {
    const base = await localBase();
    const form = new FormData();
    form.append("chunk", args.chunk, `chunk-${args.index}.wav`);
    form.append("index", String(args.index));
    form.append("total", String(args.total));
    const res = await fetch(
      `${base}/local-ai/audio-session/${encodeURIComponent(args.sessionId)}/chunk`,
      { method: "POST", body: form, signal: args.signal },
    );
    if (!res.ok) {
      const detail = await res.text().catch(() => "");
      throw new Error(`appendAudioChunk ${res.status} ${detail}`);
    }
    return (await res.json()) as AudioSessionChunkResponse;
  },

  /** Send the "now make the notes" turn. Returns the structured note
   * the renderer drops into `content_markdown`. The session is freed
   * server-side regardless of success / failure. */
  async finalizeAudioSession(
    sessionId: string,
    args: AudioSessionFinalizeRequest = {},
  ): Promise<AudioSessionFinalizeResponse> {
    const base = await localBase();
    const res = await fetch(
      `${base}/local-ai/audio-session/${encodeURIComponent(sessionId)}/finalize`,
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(args),
        // Real model inference can take a while on CPU — give it the
        // same 5-minute ceiling the cloud /ai routes use.
        signal: AbortSignal.timeout(5 * 60_000),
      },
    );
    if (!res.ok) {
      const detail = await res.text().catch(() => "");
      throw new Error(`finalizeAudioSession ${res.status} ${detail}`);
    }
    return (await res.json()) as AudioSessionFinalizeResponse;
  },
};

File: apps/api/app/routes/ai.py
from fastapi import APIRouter

from app.schemas import (
    AskRequest,
    AskResponse,
    ClassOverviewRequest,
    ClassOverviewResponse,
    FlashcardsRequest,
    FlashcardsResponse,
    NoteContext,
    QuizRequest,
    QuizResponse,
    SimpleExplainRequest,
    StudyPlanRequest,
    StudyPlanResponse,
    SummaryResponse,
)
from app.services import ai_service

router = APIRouter(prefix="/ai", tags=["ai"])


@router.post("/summarize", response_model=SummaryResponse)
async def post_summarize(req: NoteContext) -> SummaryResponse:
    data = await ai_service.summarize(title=req.title, content=req.content)
    return SummaryResponse(**data)


@router.post("/flashcards", response_model=FlashcardsResponse)
async def post_flashcards(req: FlashcardsRequest) -> FlashcardsResponse:
    data = await ai_service.flashcards(title=req.title, content=req.content, count=req.count)
    return FlashcardsResponse(**data)


@router.post("/quiz", response_model=QuizResponse)
async def post_quiz(req: QuizRequest) -> QuizResponse:
    data = await ai_service.quiz(
        title=req.title, content=req.content, count=req.count, types=req.types
    )
    return QuizResponse(**data)


@router.post("/study-plan", response_model=StudyPlanResponse)
async def post_study_plan(req: StudyPlanRequest) -> StudyPlanResponse:
    data = await ai_service.study_plan(
        goal=req.goal,
        exam_date=req.exam_date,
        notes=req.notes,
        days_available=req.days_available,
    )
    return StudyPlanResponse(**data)


@router.post("/explain-simple", response_model=SummaryResponse)
async def post_explain_simple(req: SimpleExplainRequest) -> SummaryResponse:
    data = await ai_service.explain_simple(
        title=req.title, content=req.content, audience=req.audience
    )
    return SummaryResponse(**data)


@router.post("/class-overview", response_model=ClassOverviewResponse)
async def post_class_overview(req: ClassOverviewRequest) -> ClassOverviewResponse:
    data = await ai_service.class_overview(
        class_name=req.class_name,
        class_subtitle=req.class_subtitle,
        notes=[n.model_dump() for n in req.notes],
    )
    return ClassOverviewResponse(**data)


@router.post("/ask", response_model=AskResponse)
async def post_ask(req: AskRequest) -> AskResponse:
    data = await ai_service.ask_class(
        class_name=req.class_name,
        class_subtitle=req.class_subtitle,
        recent_notes=[n.model_dump() for n in req.recent_notes],
        weak_topics=req.weak_topics,
        history=[m.model_dump() for m in req.history],
        question=req.question,
    )
    return AskResponse(**data)
